From 2a52ba9d9be0f9d04f0c7ed8b8e598204fdab558 Mon Sep 17 00:00:00 2001
From: Venkatasaiprudhvi Raj Sannidhi <vsannidhi@maxlinear.com>
Date: Mon, 2 Feb 2026 19:33:36 -0800
Subject: [PATCH] WLANRTSYS-100618 MLO: Triband MLD Support     - Extending ML
 Probe response to support 3 link AP MLD     - Refactoring the ML IE creation
 logic for Beacon, Probe, Assoc, 	  EPCS variants.

Signed-off-by: Venkatasaiprudhvi Raj Sannidhi <vsannidhi@maxlinear.com>
---
 hostapd/mxl_ctrl_iface.c       |   2 +-
 src/ap/mxl_beacon.c            |   4 +-
 src/ap/mxl_beacon.h            |   2 +-
 src/ap/mxl_ieee802_11.c        |  26 +-
 src/ap/mxl_ieee802_11_shared.c |   2 +-
 src/ap/mxl_mld.c               | 951 ++++++++++++++++-----------------
 src/ap/mxl_mld.h               |   7 +-
 src/ap/mxl_wmm.c               |   3 +-
 8 files changed, 480 insertions(+), 517 deletions(-)

diff --git a/hostapd/mxl_ctrl_iface.c b/hostapd/mxl_ctrl_iface.c
index adfbe27e2..ddeba138a 100644
--- a/hostapd/mxl_ctrl_iface.c
+++ b/hostapd/mxl_ctrl_iface.c
@@ -301,7 +301,7 @@ static int mxl_hostapd_ctrl_iface_send_epcs_req_frame(struct hostapd_data *hapd,
 	*pos++ = WLAN_ACTION_PROTECTED_EHT;
 	*pos++ = WLAN_EHT_PROTECTED_EPCS_PRIORITY_ACCESS_ENABLE_REQ;
 	*pos++ = dialog_token;
-	pos = mxl_hostapd_eid_multilink(hapd, NULL, pos, ML_EPCS, 0, false, false);
+	pos = mxl_hostapd_eid_multilink_common(hapd, pos, false, ML_EPCS);
 
 	len = pos - buf;
 
diff --git a/src/ap/mxl_beacon.c b/src/ap/mxl_beacon.c
index cef320194..ef07a5d88 100644
--- a/src/ap/mxl_beacon.c
+++ b/src/ap/mxl_beacon.c
@@ -119,10 +119,10 @@ size_t mxl_hostapd_max_chan_switch_time_ie_len(struct hostapd_data *hapd)
 		return MAX_CHANNEL_SWITCH_TIME_IE_LEN;
 }
 
-u8 * mxl_hostapd_eid_multilink_is_mlo(struct hostapd_data *hapd, u8 *pos, enum multilink_type ml_type, bool is_non_tx_ml_probe)
+u8 * mxl_hostapd_eid_multilink_is_mlo(struct hostapd_data *hapd, u8 *pos, enum multilink_type ml_type, bool include_mld_id)
 {
 	if (hapd->iconf->ieee80211be && hapd->conf->mxl_bss_conf.mlo_enable && hapd->conf->mxl_bss_conf.ml_ie_add) {
-		pos = mxl_hostapd_eid_multilink(hapd, NULL, pos, ml_type, 0, false, is_non_tx_ml_probe);
+		pos = mxl_hostapd_eid_multilink_common(hapd, pos, include_mld_id, ml_type);
 	}
 	return pos;
 }
diff --git a/src/ap/mxl_beacon.h b/src/ap/mxl_beacon.h
index 943bae17d..0d53ac16e 100644
--- a/src/ap/mxl_beacon.h
+++ b/src/ap/mxl_beacon.h
@@ -19,7 +19,7 @@
 size_t mxl_hostapd_eid_ml_probe_len(struct hostapd_data *hapd, const u8 *probe_req_mle, size_t probe_req_mle_len);
 u8 * mxl_hostapd_eid_ml_probe(struct hostapd_data *hapd, const u8 *probe_req_mle, size_t probe_req_mle_len, u8 *pos);
 size_t mxl_hostapd_max_chan_switch_time_ie_len(struct hostapd_data *hapd);
-u8 * mxl_hostapd_eid_multilink_is_mlo(struct hostapd_data *hapd, u8 *pos, enum multilink_type ml_type, bool is_non_tx_ml_probe);
+u8 * mxl_hostapd_eid_multilink_is_mlo(struct hostapd_data *hapd, u8 *pos, enum multilink_type ml_type, bool include_mld_id);
 bool mxl_check_ml_probe_type_address_check(struct hostapd_data *hapd, const struct ieee80211_mgmt *mgmt);
 size_t mxl_hostapd_multilink_ie_len_is_mlo(struct hostapd_data *hapd, enum multilink_type ml_type);
 int mxl_hostap_ml_get_non_tx_bss_idx_from_mlid(struct hostapd_data *hapd,  const u8 *probe_req_mle, size_t probe_req_mle_len);
diff --git a/src/ap/mxl_ieee802_11.c b/src/ap/mxl_ieee802_11.c
index 4b2975e94..f67d46b4d 100644
--- a/src/ap/mxl_ieee802_11.c
+++ b/src/ap/mxl_ieee802_11.c
@@ -364,9 +364,31 @@ u16 mxl_check_assoc_ie_ml_ie_validate(struct hostapd_data *hapd, struct sta_info
 
 void mxl_auth_hostapd_eid_multilink(struct hostapd_data *hapd, struct sta_info *sta, u8 *eid)
 {
-	if (sta && mxl_is_mlo_enabled(hapd->conf) && mxl_mlo_is_ml_sta(sta)) {
-		mxl_hostapd_eid_multilink(hapd, sta, eid, ML_BASIC, WLAN_STATUS_SUCCESS, true, false);
+	size_t auth_len = mxl_ml_auth_len(hapd, sta);
+	struct wpabuf *buf = NULL;
+
+	if (auth_len == 0) {
+		wpa_printf(MSG_DEBUG, "MXL_MLO: ML Auth IE need not be added");
+		return;
 	}
+
+	buf = wpabuf_alloc(auth_len);
+
+	if (!buf) {
+		wpa_printf(MSG_ERROR, "MXL_MLO: Failed to allocate buffer to build ML Auth Response IE");
+		return;
+	}
+
+	wpabuf_put_u8(buf, WLAN_EID_EXTENSION);
+	wpabuf_put_u8(buf, MULTILINK_IE_LEN_AUTH);
+	wpabuf_put_u8(buf, WLAN_EID_EXT_MULTI_LINK);
+	wpabuf_put_le16(buf, MULTI_LINK_CONTROL_TYPE_BASIC);
+	wpabuf_put_u8(buf, COMMON_INFO_LEN_AUTH);
+	wpabuf_put_data(buf, hapd->conf->mxl_bss_conf.ap_mld_mac, ETH_ALEN);
+
+	os_memcpy(eid, wpabuf_head(buf), wpabuf_len(buf));
+	eid += wpabuf_len(buf);
+	wpabuf_free(buf);
 }
 
 int mxl_mlo_sae_prepare_commit_pt(struct hostapd_data *hapd, struct sta_info *sta, const struct sae_pt *pt, const struct sae_pk *pk)
diff --git a/src/ap/mxl_ieee802_11_shared.c b/src/ap/mxl_ieee802_11_shared.c
index 2394e9b64..cd753c9f1 100644
--- a/src/ap/mxl_ieee802_11_shared.c
+++ b/src/ap/mxl_ieee802_11_shared.c
@@ -45,7 +45,7 @@ u8 * mxl_hostapd_eid_ml_mbssid(struct hostapd_data *hapd, struct hostapd_data *n
 		}
 
 		if (add_mlo_ie) {
-			pos = mxl_hostapd_eid_multilink(non_tx_hapd, NULL, pos, ML_BASIC, 0, false, false);
+			pos = mxl_hostapd_eid_multilink_common(non_tx_hapd, pos, false, ML_BASIC);
 		}
 
 		/* TO-DO: Workaround to add WMM and Vendor Specific IE to get pass for Wi-Fi 7
diff --git a/src/ap/mxl_mld.c b/src/ap/mxl_mld.c
index 3d93795e9..beec35905 100644
--- a/src/ap/mxl_mld.c
+++ b/src/ap/mxl_mld.c
@@ -51,16 +51,6 @@ static void mxl_mlo_set_link_ids(struct p_ap_mld *mld)
 	}
 }
 
-#ifdef CONFIG_EPCS_MXL
-static u8 * mxl_hostapd_eid_mld_staprof_epcs(struct hostapd_data *hapd, u8 *eid)
-{
-	eid = mxl_hostapd_eid_wmm_epcs(hapd, eid);
-	eid = hostapd_eid_he_mu_edca_parameter_set(hapd, eid);
-
-	return eid;
-}
-#endif /* CONFIG_EPCS_MXL */
-
 static int mxl_get_valid_mld_id(struct hostapd_data *hapd, struct p_ap_mld *mld)
 {
 	struct hapd_interfaces *ifaces = hapd->iface->interfaces;
@@ -1096,15 +1086,319 @@ end:
 	return ret_buf;
 }
 
+static void mxl_ml_fragment_per_sta_profile(struct wpabuf *buf, u8 *per_sta_cap,
+												u16 per_sta_cap_len,
+												u16 total_len,
+												size_t sta_info_len)
+{
+	u8 *ptr = per_sta_cap;
+	size_t len = per_sta_cap_len;
+	size_t slice_len = MAX_IE_LEN - sta_info_len;
+
+	wpa_printf(MSG_DEBUG, "MXL_MLO: Fragmenting per STA profile, total len %u, per_sta_cap_len %u",
+																		total_len, per_sta_cap_len);
+
+	wpabuf_put_data(buf, ptr, slice_len);
+	len -= slice_len;
+	ptr += slice_len;
+
+	while (len) {
+		wpa_printf(MSG_DEBUG, "MXL_MLO: Fragmenting per STA profile, remaining len %zu", len);
+		if (len <= MAX_IE_LEN)
+			slice_len = len;
+		else
+			slice_len = MAX_IE_LEN;
+
+		wpabuf_put_u8(buf, EHT_ML_SUB_ELEM_FRAGMENT);
+		wpabuf_put_u8(buf, slice_len);
+		wpabuf_put_data(buf, ptr, slice_len);
+		len -= slice_len;
+		ptr += slice_len;
+	}
+}
+
+static u8 * mxl_ml_ie_prepare(struct wpabuf *buf, u8 *pos)
+{
+	size_t slice_len = 0;
+	size_t len = wpabuf_len(buf);
+	const u8 *ptr = wpabuf_head(buf);
+
+	if (!ptr) {
+		wpa_printf(MSG_ERROR, "MXL_MLO: %s - wpabuf_head returned NULL!", __func__);
+		return pos;
+	}
+
+	if (len <= 254)
+		slice_len = len;
+	else
+		slice_len = 254;
+
+	*pos++ = WLAN_EID_EXTENSION;
+	*pos++ = slice_len + 1;
+	*pos++ = WLAN_EID_EXT_MULTI_LINK;
+	os_memcpy(pos, ptr, slice_len);
+
+	ptr += slice_len;
+	pos += slice_len;
+	len -= slice_len;
+
+	/* Fragment the Multi-Link element, if needed */
+	while (len) {
+		if (len <= 255)
+			slice_len = len;
+		else
+			slice_len = 255;
+
+		*pos++ = WLAN_EID_FRAGMENT;
+		*pos++ = slice_len;
+		os_memcpy(pos, ptr, slice_len);
+
+		ptr += slice_len;
+		pos += slice_len;
+		len -= slice_len;
+	}
+	return pos;
+}
+
+static u16 mxl_ml_build_control_field(struct hostapd_data *hapd,
+				       bool include_mld_id, bool include_msdi,
+				       enum multilink_type ml_type)
+{
+	u16 control = 0;
+
+#ifdef CONFIG_EPCS_MXL
+	if (ml_type == ML_EPCS)
+		control |= MULTI_LINK_CONTROL_TYPE_PRIOR_ACCESS;
+	else
+#endif /* CONFIG_EPCS_MXL */
+		control |= (MULTI_LINK_CONTROL_TYPE_BASIC|
+				MLO_LINK_ID_INFO_PRESENT |
+				MLO_BSS_PARAM_CHANGE_COUNT_PRESENT |
+				MLO_EML_CAPAB_PRESENT |
+				MLO_MLD_CAPAB_PRESENT);
+
+	if (include_msdi)
+		control |= MLO_MSDI_PRESENT;
+
+	if (include_mld_id)
+		control |= MLO_MLD_ID_PRESENT;
+
+	return control;
+}
+
+static u8 mxl_ml_get_ap_mld_common_info_len(struct hostapd_data *hapd,
+										 bool include_mld_id,
+										 bool include_msdi,
+										 enum multilink_type ml_type)
+{
+	u8 common_info_len = 0;
+
+#ifdef CONFIG_EPCS_MXL
+	if (ml_type == ML_EPCS)
+		common_info_len = COMMON_INFO_LEN_AUTH;
+	else
+#endif /* CONFIG_EPCS_MXL */
+		common_info_len = COMMON_INFO_LEN;
+
+	if (include_msdi)
+		common_info_len += MULTILINK_MSD_LEN;
+
+	if (include_mld_id)
+		common_info_len ++;
+
+	wpa_printf(MSG_DEBUG, "MXL_MLO: common_info_len %d for iface %s",
+									common_info_len, hapd->conf->iface);
+	return common_info_len;
+}
+
+static void mxl_ml_build_common_info(struct hostapd_data *hapd,
+									struct wpabuf *buf,
+									bool include_mld_id,
+									bool include_msdi,
+									enum multilink_type ml_type)
+{
+	bool mld_mediumsync_present = hapd->conf->mxl_bss_conf.mld_mediumsync_present;
+	u8 mld_num_simul_links = mxl_ml_get_num_of_sim_links(hapd);
+	u8 bpcc = hapd->conf->mxl_bss_conf.mld_bpcc_count;
+	u8 common_info_len = 0;
+	u16 msd_info = 0;
+	u16 eml_capab = 0;
+	u16 mld_capab = 0;
+
+	/* Common Info Length */
+	common_info_len = mxl_ml_get_ap_mld_common_info_len(hapd, include_mld_id, include_msdi, ml_type);
+	wpabuf_put_u8(buf, common_info_len);
+
+	/* MLD MAC Address */
+	wpabuf_put_data(buf, hapd->conf->mxl_bss_conf.ap_mld_mac, ETH_ALEN);
+
+#ifdef CONFIG_EPCS_MXL
+	if (ml_type == ML_EPCS)
+		return;
+#endif /* CONFIG_EPCS_MXL */
+
+	/* Link ID */
+	wpabuf_put_u8(buf, hapd->conf->mxl_bss_conf.link_id);
+
+	/* BSS Parameter Change Count */
+	wpabuf_put_u8(buf, bpcc);
+
+	/* Medium Sync Delay Info */
+	if (mld_mediumsync_present) {
+		msd_info = hapd->conf->mxl_bss_conf.mld_mediumsync_duration;
+		msd_info |= hapd->conf->mxl_bss_conf.mld_mediumsync_ofdmedthresh << MLO_MEDIUM_SYNC_OFDMED_THRESH_START_POSITION;
+		msd_info |= hapd->conf->mxl_bss_conf.mld_mediumsync_maxtxop << MLO_MEDIUM_SYNC_MAX_TXOP_START_POSITION;
+		wpabuf_put_le16(buf, msd_info);
+	}
+
+	/* Fill EML Capab */
+	if (hapd->iconf->mxl_conf.mlo_supported == MULTILINK_EMLSR)
+		eml_capab |= MLO_EMLSR;
+	else if (hapd->iconf->mxl_conf.mlo_supported == MULTILINK_EMLMR)
+		eml_capab |= MLO_EMLMR;
+	eml_capab |= hapd->conf->mxl_bss_conf.eml_transition_timeout << MLO_TRANSITION_TIMEOUT_MASK;
+	wpabuf_put_le16(buf, eml_capab);
+
+	/* Fill MLD Capab */
+	mld_capab |= mld_num_simul_links;
+	if (hapd->conf->mxl_bss_conf.mlo_t2lm_support)
+		mld_capab |= MLO_T2LM_NEGOTIATION_SUPPORT_MAP_ALL_TID_TO_SAME_LINK;
+	wpabuf_put_le16(buf, mld_capab);
+
+	/* MLD ID */
+	if (include_mld_id)
+		wpabuf_put_u8(buf, hapd->conf->bss_idx - 1); /* BSSID IDX */
+}
+
+static void mxl_ml_build_sta_control_field(struct wpabuf *buf,
+									   struct hostapd_data *lhapd,
+									   enum multilink_type ml_type)
+{
+	u16 sta_ctrl = 0;
+
+#ifdef CONFIG_EPCS_MXL
+	if (ml_type == ML_EPCS) {
+		sta_ctrl = lhapd->conf->mxl_bss_conf.link_id;
+		sta_ctrl = host_to_le16(sta_ctrl);
+		goto out;
+	}
+#endif /* CONFIG_EPCS_MXL */
+	/* CSA Multilink case, beacons to add partial profile with CSA/ECSA/MCST IE */
+	if (ml_type != ML_BASIC)
+		sta_ctrl |= (MLO_LINK_INFO_COMPLETE_PROFILE);
+	if (ml_type == ML_ASSOC)
+		sta_ctrl |= (MLO_LINK_INFO_BSS_PARAM_CHANGE_CNT_PRESENT);
+
+	sta_ctrl |= (lhapd->conf->mxl_bss_conf.link_id |
+				MLO_LINK_INFO_MACADDR_PRESENT | MLO_LINK_INFO_BI_PRESENT |
+				MLO_LINK_INFO_TSF_OFFSET_PRESENT | MLO_LINK_INFO_DTIM_INFO_PRESENT);
+
+out:
+	wpabuf_put_le16(buf, sta_ctrl);
+}
+
+static size_t mxl_ml_build_sta_info(struct wpabuf *buf,
+							 struct hostapd_data *lhapd,
+							 struct hostapd_data *hapd,
+							 enum multilink_type ml_type)
+{
+	u8 *sta_info_len = NULL;
+	u64 tsf = 0;
+
+	sta_info_len = wpabuf_put(buf, 1); /* Placeholder to update the length later */
+	wpabuf_put_data(buf, lhapd->own_addr, ETH_ALEN);
+	wpabuf_put_le16(buf, lhapd->iconf->beacon_int);
+#ifdef CONFIG_WIFI_CERTIFICATION
+	if((hapd->iface->current_mode->mode == HOSTAPD_MODE_IEEE80211B ||
+			hapd->iface->current_mode->mode == HOSTAPD_MODE_IEEE80211G) ||
+			(lhapd->iface->current_mode->mode == HOSTAPD_MODE_IEEE80211B ||
+			lhapd->iface->current_mode->mode == HOSTAPD_MODE_IEEE80211G))
+		tsf = hapd->iconf->mxl_conf.eht_mld_tsf_diff;
+#endif /* CONFIG_WIFI_CERTIFICATION */
+	wpabuf_put_le64(buf, tsf);
+	wpabuf_put_u8(buf, 0); /* DTIM Count */
+	wpabuf_put_u8(buf, lhapd->conf->dtim_period);
+
+	if (ml_type == ML_ASSOC) {
+		*sta_info_len = ML_PER_STA_INFO_LEN;
+		wpabuf_put_u8(buf, lhapd->conf->mxl_bss_conf.mld_bpcc_count);
+	} else {
+		/* No BPCC in other frame types, adjust the length accordingly */
+		*sta_info_len = ML_PER_STA_INFO_LEN - 1;
+	}
+
+	return *sta_info_len;
+}
+
+static bool mxl_ml_is_per_sta_profile_required(struct hostapd_data *hapd,
+												enum multilink_type ml_type)
+{
+	/* If CSA IEs need to be included, include per STA profile */
+	if ((ml_type != ML_BASIC) ||
+	    (hapd->mxl_data.ml_ie_max_chan_sw_time_ie_add ||
+	     hapd->mxl_data.ml_ie_max_chan_sw_time_ie_len_add)) {
+			return true;
+	}
+	return false;
+}
+
+static u8 * mxl_hostapd_eid_mld_staprof_bcn(struct hostapd_data *hapd, u8 *eid, u8 *fw_ie_len)
+{
+	u8 fw_ie_len_calc = 0;
+
+	if (!hapd || !eid || !fw_ie_len)
+		return NULL;
+
+	*fw_ie_len = 0;
+
+	wpa_printf(MSG_INFO, "MXL_MLO: mld_staprof_bcn: ml_ie_max_chan_sw_time_ie_add %d\
+			      ml_ie_csa_ie_len_add %d ml_ie_csw_ie_len_add %d ml_ie_max_chan_sw_time_ie_len_add %d",
+			      hapd->mxl_data.ml_ie_max_chan_sw_time_ie_add, hapd->mxl_data.ml_ie_csa_ie_len_add,
+			      hapd->mxl_data.ml_ie_csw_ie_len_add, hapd->mxl_data.ml_ie_max_chan_sw_time_ie_len_add);
+
+	if (hapd->mxl_data.ml_ie_max_chan_sw_time_ie_add &&
+	    !hapd->mxl_data.ml_ie_max_chan_sw_time_ie_len_add) {
+		eid = mxl_hostapd_eid_max_chan_sw_time(hapd, eid);
+	}
+
+	if (hapd->mxl_data.ml_ie_csa_ie_len_add) {
+		fw_ie_len_calc += CSA_IE_LEN;
+	}
+
+	if (hapd->mxl_data.ml_ie_csw_ie_len_add) {
+		fw_ie_len_calc += CSW_IE_LEN;
+	}
+
+	if (hapd->mxl_data.bw_indication_ie_len_add) {
+		if (hapd->iconf->punct_bitmap) {
+			fw_ie_len_calc += BWI_IE_LEN;
+		} else {
+			if (hostapd_get_oper_chwidth(hapd->iface->conf) == EHT_OPER_CHANWIDTH_320) {
+				fw_ie_len_calc += BWI_IE_NO_DSB_LEN;
+			}
+		}
+	}
+
+	if ((hapd->mxl_data.ml_ie_max_chan_sw_time_ie_len_add) &&
+            (!hapd->mxl_data.ml_ie_max_chan_sw_time_ie_add)) {
+		fw_ie_len_calc += MAX_CHAN_SWITCH_TIME_IE_LEN;
+	}
+
+	*fw_ie_len = fw_ie_len_calc;
+
+	wpa_printf(MSG_DEBUG, "MXL_MLO: mld_staprof_bcn: fw_ie_len %d", (*fw_ie_len));
+
+	return eid;
+}
+
 static u8 * mxl_hostapd_eid_mld_staprof_probe(struct hostapd_data *hapd, u8 *eid)
 {
 	struct hostapd_hw_modes *mode;
 
-	mode = hapd->iface->current_mode;
+	if (!hapd || !eid)
+		return NULL;
 
-	/* capability info */
-	WPA_PUT_LE16(eid, hostapd_own_capab_info(hapd));
-	eid += sizeof(le16);
+	mode = hapd->iface->current_mode;
 
 	/* Supported rates */
 	eid = hostapd_eid_supp_rates(hapd, eid);
@@ -1222,147 +1516,143 @@ done:
 	return eid;
 }
 
-/*
- * mxl_hostapd_eid_mld_frag - API to adopt fragments for ML IE/per sta profile
- * if MultiLink IE exceeds max IE length 255, because of the per sta
- * profile size.
- *
- * ml_ie_data_start - MultiLink IE data start
- * common_info_len - length of common info
- * per_sta_prof_total_len - Total per sta profile length that need to be
- *				packed as fragments
- *
- * TODO - rework the fragmentation logic to avoid in-place buffer updation
- */
-u8 * mxl_hostapd_eid_mld_frag(u8 *ml_ie_data_start,
-			  u8 common_info_len, u16 per_sta_prof_total_len)
-{
-	u8 ml_ie_data_len;
-	u8 *eid = ml_ie_data_start, *buf_pos, *ml_frag_len;
-	u8 *per_sta_prof_buf;
-	u16 per_sta_cur_frag_len, per_sta_rem_frag_len;
-	u16 per_sta_rem_total = per_sta_prof_total_len;
-
-	per_sta_prof_buf = os_malloc(PER_STA_PROFILE_LENGTH * sizeof(u8));
-	if (!per_sta_prof_buf) {
-		wpa_printf(MSG_ERROR, "os_malloc fails for per_sta_prof_buf");
+#ifdef CONFIG_EPCS_MXL
+static u8 * mxl_hostapd_eid_mld_staprof_epcs(struct hostapd_data *hapd, u8 *eid)
+{
+	if (!hapd || !eid)
 		return NULL;
-	}
-
-	/* Length of Multilink IE except per sta prof subelement
-	 * ext eid(107) + multilink control len + common info len
-	 */
-	ml_ie_data_len = (1 + MULTILINK_CTRL_LEN + common_info_len);
 
-	/* per sta profile fragment length that can be packed in 1st ML IE */
-	per_sta_cur_frag_len = (MAX_IE_LEN - ml_ie_data_len - EID_HDR_LEN);
-
-	if (per_sta_prof_total_len < per_sta_cur_frag_len) {
-		wpa_printf(MSG_ERROR, "ml ie frag tx error: per_sta_prof_total_len %d, per_sta_cur_frag_len %d",
-				per_sta_prof_total_len, per_sta_cur_frag_len);
-		goto frag_done;
-	}
+	eid = mxl_hostapd_eid_wmm_epcs(hapd, eid);
+	eid = hostapd_eid_he_mu_edca_parameter_set(hapd, eid);
 
-	/* per sta profile data remaining */
-	per_sta_rem_total = (per_sta_prof_total_len - per_sta_cur_frag_len);
+	return eid;
+}
+#endif /* CONFIG_EPCS_MXL */
 
-	/* per sta profile remaining fragment to be packed in start of 2nd ML fragment IE */
-	per_sta_rem_frag_len = (MAX_IE_LEN - per_sta_cur_frag_len);
+static u8 * mxl_hostapd_eid_mld_staprof(struct hostapd_data *hapd, u8 *buf, u16 *ie_len,
+										enum multilink_type ml_type, u8 *fw_ie_len)
+{
+	u8 *eid = buf;
+	le16 capab_info = 0;
 
-	/* Check if there is a need for per sta profile fragmentation */
-	if (per_sta_rem_frag_len > per_sta_rem_total)
-		per_sta_rem_frag_len = per_sta_rem_total;
+	if (!hapd || !buf || !ie_len)
+		return NULL;
 
-	eid += (ml_ie_data_len + EID_HDR_LEN + per_sta_cur_frag_len);
-	memcpy_s(per_sta_prof_buf, per_sta_rem_total, eid, per_sta_rem_total);
-	buf_pos = per_sta_prof_buf;
+	/* Set initial len */
+	*ie_len = 0;
 
-	while (per_sta_rem_frag_len) {
+	/* capability info - common for all types */
+	capab_info = host_to_le16(hostapd_own_capab_info(hapd));
+	memcpy_s(eid, sizeof(capab_info), &capab_info, sizeof(capab_info));
+	eid += sizeof(capab_info);
 
-		*eid++ = WLAN_EID_FRAGMENT;
-		ml_frag_len = eid++;
+	/* Build ML type-specific per sta profile content */
+	switch (ml_type) {
+	case ML_BASIC:
+		eid = mxl_hostapd_eid_mld_staprof_bcn(hapd, eid, fw_ie_len);
+		break;
 
-		memcpy_s(eid, per_sta_rem_frag_len, buf_pos, per_sta_rem_frag_len);
-		eid += per_sta_rem_frag_len;
-		buf_pos += per_sta_rem_frag_len;
-		per_sta_rem_total -= per_sta_rem_frag_len;
+	case ML_PROBE:
+		eid = mxl_hostapd_eid_mld_staprof_probe(hapd, eid);
+		break;
 
-		/* Case 1: Additional per sta profile fragments required */
-		if (per_sta_rem_total > (MAX_IE_LEN - EID_HDR_LEN - per_sta_rem_frag_len)) {
-			per_sta_cur_frag_len = (MAX_IE_LEN - EID_HDR_LEN - per_sta_rem_frag_len);
-			per_sta_rem_frag_len = (MAX_IE_LEN - per_sta_cur_frag_len);
-			*ml_frag_len = MAX_IE_LEN;
-		/* Case 2: current per sta profile fragment with length less than 255 is enough */
-		} else if (per_sta_rem_total) {
-			per_sta_cur_frag_len = per_sta_rem_total;
-			*ml_frag_len = (per_sta_rem_frag_len + EID_HDR_LEN + per_sta_cur_frag_len);
-			per_sta_rem_frag_len = 0;
-		/* Case 3: No more per sta profile fragments required, current current ML frag length */
-		} else {
-			*ml_frag_len = per_sta_rem_frag_len;
-			goto frag_done;
-		}
+#ifdef CONFIG_EPCS_MXL
+	case ML_EPCS:
+		eid = mxl_hostapd_eid_mld_staprof_epcs(hapd, eid);
+		break;
+#endif /* CONFIG_EPCS_MXL */
+	default:
+		wpa_printf(MSG_ERROR, "MXL_MLO: Invalid ml_type %d", ml_type);
+		return NULL;
+	}
 
-		*eid++ = PER_STA_FRAG_SUB_ELEM_ID;
-		*eid++ = (per_sta_cur_frag_len + per_sta_rem_frag_len);
+	if (!eid)
+		return NULL;
 
-		memcpy_s(eid, per_sta_cur_frag_len, buf_pos, per_sta_cur_frag_len);
-		eid += per_sta_cur_frag_len;
-		buf_pos += per_sta_cur_frag_len;
-		per_sta_rem_total -= per_sta_cur_frag_len;
-	}
+	*ie_len = eid - buf;
 
-frag_done:
-	os_free(per_sta_prof_buf);
 	return eid;
 }
 
-static u8 * mxl_hostapd_eid_mld_staprof_bcn(struct hostapd_data *hapd, u8 *eid, u8 *fw_ie_len)
+u8 * mxl_hostapd_eid_multilink_common(struct hostapd_data *hapd, u8 *eid,
+								bool include_mld_id, enum multilink_type ml_type)
 {
-	u8 ie_len = 0;
-	struct hostapd_data *sibling_hapd = hapd->mxl_data.sibling_hapd[LINK1]; /* TODO: WLANRTSYS-95653 - Need to handle during the CSA task */
+	bool include_msdi = hapd->conf->mxl_bss_conf.mld_mediumsync_present;
+	u8 *pos = eid;
+	u16 control = 0;
+	struct wpabuf *buf;
+	struct hostapd_data *lhapd;
+	struct p_ap_mld *mld = mxl_ml_get_ap_mld(hapd);
 
-	wpa_printf(MSG_DEBUG, "MXL_MLO: mld_staprof_bcn: ml_ie_max_chan_sw_time_ie_add %d\
-			      ml_ie_csa_ie_len_add %d ml_ie_csw_ie_len_add %d ml_ie_max_chan_sw_time_ie_len_add %d",
-			      hapd->mxl_data.ml_ie_max_chan_sw_time_ie_add, hapd->mxl_data.ml_ie_csa_ie_len_add,
-			      hapd->mxl_data.ml_ie_csw_ie_len_add, hapd->mxl_data.ml_ie_max_chan_sw_time_ie_len_add);
+	if (!mld) {
+		wpa_printf(MSG_ERROR, "MXL_MLO: No MLD found for BSS %s", hapd->conf->iface);
+		return NULL;
+	}
 
-	WPA_PUT_LE16(eid, hostapd_own_capab_info(sibling_hapd));
-        eid += sizeof(u16);
+	/*
+	 * As the Multi-Link element can exceed the size of 255 bytes need to
+	 * first build it and then handle fragmentation.
+	 */
+	buf = wpabuf_alloc(EHT_ML_MAX_STA_PROF_LEN);
+	if (!buf)
+		return pos;
 
-	if (hapd->mxl_data.ml_ie_max_chan_sw_time_ie_add &&
-	    !hapd->mxl_data.ml_ie_max_chan_sw_time_ie_len_add) {
-		eid = mxl_hostapd_eid_max_chan_sw_time(hapd, eid);
-	}
+	/* Multi-Link Control field */
+	control = mxl_ml_build_control_field(hapd, include_mld_id, include_msdi, ml_type);
+	wpabuf_put_le16(buf, control);
 
-	if (hapd->mxl_data.ml_ie_csa_ie_len_add) {
-		ie_len += CSA_IE_LEN;
-	}
+	/* Build Common Info section */
+	mxl_ml_build_common_info(hapd, buf, include_mld_id, include_msdi, ml_type);
+	wpa_printf(MSG_DEBUG, "MXL_MLO: ml_type %d, MCST add host %d MCST add fw %d",
+				ml_type, hapd->mxl_data.ml_ie_max_chan_sw_time_ie_add,
+				hapd->mxl_data.ml_ie_max_chan_sw_time_ie_len_add);
 
-	if (hapd->mxl_data.ml_ie_csw_ie_len_add) {
-		ie_len += CSW_IE_LEN;
-	}
+	if (mxl_ml_is_per_sta_profile_required(hapd, ml_type)) {
+		/* Build per STA profiles */
+		mxl_mld_for_each_sibling_link(lhapd, hapd) {
+			u8 fw_ie_len = 0;
+			u8 *link_info_len = NULL;
+			u8 per_sta_cap_buf[EHT_ML_MAX_STA_PROF_LEN];
+			u16 per_sta_cap_len = 0;
+			size_t total_len = 0;
+			size_t sta_info_len = 0;
+
+			/* Build Per-STA profile */
+			mxl_hostapd_eid_mld_staprof(lhapd, per_sta_cap_buf, &per_sta_cap_len, ml_type, &fw_ie_len);
+			wpa_printf(MSG_DEBUG, "MXL_MLO: fw ie len %d!!!", fw_ie_len);
+
+			if (!per_sta_cap_len) {
+				wpa_printf(MSG_ERROR, "MXL_MLO: Failed to build per-STA profile for link %s", lhapd->conf->iface);
+				continue;
+			}
+			wpabuf_put_u8(buf, WLAN_EID_SUBELEMENT_MLO_PER_STA_PROFILE);
 
-	if (hapd->mxl_data.bw_indication_ie_len_add) {
-		if (hapd->iconf->punct_bitmap) {
-			ie_len += BWI_IE_LEN;
-		} else {
-			if (hostapd_get_oper_chwidth(hapd->iface->conf) == EHT_OPER_CHANWIDTH_320) {
-				ie_len += BWI_IE_NO_DSB_LEN;
+			link_info_len = wpabuf_put(buf, 1); /* Placeholder to update the length later */
+
+			/* STA control field */
+			mxl_ml_build_sta_control_field(buf, lhapd, ml_type);
+
+			/* STA Info */
+			sta_info_len = mxl_ml_build_sta_info(buf, lhapd, hapd, ml_type);
+
+			/* Prepare per-STA Profile subelement length */
+			total_len = LINK_INFO_STA_CTRL_LEN + sta_info_len + per_sta_cap_len + fw_ie_len;
+
+			if (total_len <= MAX_IE_LEN) {
+				*link_info_len = total_len;
+				wpabuf_put_data(buf, per_sta_cap_buf, per_sta_cap_len);
+			} else {
+				*link_info_len = MAX_IE_LEN;
+				mxl_ml_fragment_per_sta_profile(buf, per_sta_cap_buf, per_sta_cap_len, total_len, sta_info_len);
 			}
 		}
 	}
 
-	if ((hapd->mxl_data.ml_ie_max_chan_sw_time_ie_len_add) &&
-            (!hapd->mxl_data.ml_ie_max_chan_sw_time_ie_add)) {
-		ie_len += MAX_CHAN_SWITCH_TIME_IE_LEN;
-	}
-
-	*fw_ie_len = ie_len;
+	/* Prepare the Multi-Link IE, fragment if needed */
+	pos = mxl_ml_ie_prepare(buf, pos);
 
-	wpa_printf(MSG_DEBUG, "MXL_MLO: mld_staprof_bcn: fw_ie_len %d", (*fw_ie_len));
-
-	return eid;
+	wpabuf_free(buf);
+	return pos;
 }
 
 u8 * mxl_ml_eid_multilink_assoc(struct hostapd_data *hapd, struct sta_info *sta, u8 *eid)
@@ -1371,24 +1661,9 @@ u8 * mxl_ml_eid_multilink_assoc(struct hostapd_data *hapd, struct sta_info *sta,
 	struct sta_info *lsta;
 	struct wpabuf *buf;
 	u8 *pos = eid;
-	const u8 *ptr;
-	size_t len, slice_len;
 	u8 link_id;
-	u8 common_info_len;
 	u16 control = 0;
-	u16 msd_info = 0;
-	u16 eml_capab = 0;
-	u16 mld_capab = 0;
-	u8 mld_num_simul_links = mxl_ml_get_num_of_sim_links(hapd);
-	bool mld_mediumsync_present = hapd->conf->mxl_bss_conf.mld_mediumsync_present;
-	u8 bpcc = hapd->conf->mxl_bss_conf.mld_bpcc_count;
-
-	size_t sta_info_len = 0;
-	size_t total_len = 0;
-	u8 *per_sta_cap = NULL;
-	size_t per_sta_cap_len = 0;
-	u16 sta_ctrl = 0;
-	u64 tsf = 0;
+	bool include_msdi = hapd->conf->mxl_bss_conf.mld_mediumsync_present;
 
 	/*
 	 * As the Multi-Link element can exceed the size of 255 bytes need to
@@ -1399,54 +1674,19 @@ u8 * mxl_ml_eid_multilink_assoc(struct hostapd_data *hapd, struct sta_info *sta,
 		return pos;
 
 	/* Multi-Link Control field */
-	control |= MULTI_LINK_CONTROL_TYPE_BASIC;
-	control |= (MLO_LINK_ID_INFO_PRESENT |
-		MLO_BSS_PARAM_CHANGE_COUNT_PRESENT |
-		MLO_MLD_CAPAB_PRESENT);
-	control |= MLO_EML_CAPAB_PRESENT;
-	if (mld_mediumsync_present)
-		control |= MLO_MSDI_PRESENT;
-
-	/* Set the basic Multi-Link common information */
-	common_info_len = COMMON_INFO_LEN;
-
-	if (mld_mediumsync_present)
-		common_info_len += MULTILINK_MSD_LEN;
-
+	control = mxl_ml_build_control_field(hapd, false, include_msdi, ML_ASSOC);
 	wpabuf_put_le16(buf, control);
 
-	wpabuf_put_u8(buf, common_info_len);
-
-	/* MLD MAC Address */
-	wpabuf_put_data(buf, hapd->conf->mxl_bss_conf.ap_mld_mac, ETH_ALEN);
-
-	/* Link ID */
-	wpabuf_put_u8(buf, hapd->conf->mxl_bss_conf.link_id);
-
-	wpabuf_put_u8(buf, bpcc);
-
-	if (mld_mediumsync_present) {
-		msd_info = hapd->conf->mxl_bss_conf.mld_mediumsync_duration;
-		msd_info |= hapd->conf->mxl_bss_conf.mld_mediumsync_ofdmedthresh << MLO_MEDIUM_SYNC_OFDMED_THRESH_START_POSITION;
-		msd_info |= hapd->conf->mxl_bss_conf.mld_mediumsync_maxtxop << MLO_MEDIUM_SYNC_MAX_TXOP_START_POSITION;
-		wpabuf_put_le16(buf, msd_info);
-	}
-
-	/* Fill EML Capab */
-	if (hapd->iconf->mxl_conf.mlo_supported == MULTILINK_EMLSR)
-		eml_capab |= MLO_EMLSR;
-	else if (hapd->iconf->mxl_conf.mlo_supported == MULTILINK_EMLMR)
-		eml_capab |= MLO_EMLMR;
-	eml_capab |= hapd->conf->mxl_bss_conf.eml_transition_timeout << MLO_TRANSITION_TIMEOUT_MASK;
-	wpabuf_put_le16(buf, eml_capab);
-
-	/* Fill MLD Capab */
-	mld_capab |= mld_num_simul_links;
-	if (hapd->conf->mxl_bss_conf.mlo_t2lm_support)
-		mld_capab |= MLO_T2LM_NEGOTIATION_SUPPORT_MAP_ALL_TID_TO_SAME_LINK;
-	wpabuf_put_le16(buf, mld_capab);
+	/* Build Common Info section */
+	mxl_ml_build_common_info(hapd, buf, false, include_msdi, ML_ASSOC);
 
 	mxl_mld_for_each_sibling_sta(lsta, sta) {
+		size_t total_len = 0;
+		size_t per_sta_cap_len = 0;
+		size_t sta_info_len = 0;
+		u8 *per_sta_cap = NULL;
+		u8 *link_info_len = NULL;
+
 		link_id = lsta->mxl_sta_info.link_id;
 		lhapd = mxl_ml_get_bss_by_link_id(hapd, link_id);
 		if (!lhapd) {
@@ -1458,342 +1698,45 @@ u8 * mxl_ml_eid_multilink_assoc(struct hostapd_data *hapd, struct sta_info *sta,
 
 		per_sta_cap = lsta->mxl_sta_info.resp_per_sta_profile;
 		per_sta_cap_len = lsta->mxl_sta_info.resp_per_sta_profile_len;
-	
-		/* Prepare next per-Sta info */
-		sta_info_len = LINK_INFO_STA_CTRL_LEN + ML_PER_STA_INFO_LEN;
-		total_len = 0;
-		sta_ctrl = 0;
-		tsf = 0;
 
-		/* Prepare per-STA Profile subelement length */
-		total_len = sta_info_len + per_sta_cap_len;
+		if (!per_sta_cap){
+			wpa_printf(MSG_ERROR,
+				   "MXL MLO: Couldn't find per-STA profile for link id %d - skip it",
+				   link_id);
+			continue;
+		}
 
 		/* Per-STA Profile subelement */
 		wpabuf_put_u8(buf, WLAN_EID_SUBELEMENT_MLO_PER_STA_PROFILE);
 
-		if (total_len <= 255)
-			wpabuf_put_u8(buf, total_len);
-		else
-			wpabuf_put_u8(buf, 255);
+		link_info_len = wpabuf_put(buf, 1); /* Placeholder to update the length later */
 
 		/* STA Control */
-		sta_ctrl |= (lhapd->conf->mxl_bss_conf.link_id | MLO_LINK_INFO_COMPLETE_PROFILE |
-			MLO_LINK_INFO_MACADDR_PRESENT | MLO_LINK_INFO_BI_PRESENT |
-			MLO_LINK_INFO_TSF_OFFSET_PRESENT | MLO_LINK_INFO_DTIM_INFO_PRESENT |
-			MLO_LINK_INFO_BSS_PARAM_CHANGE_CNT_PRESENT);
-
-		wpabuf_put_le16(buf, sta_ctrl);
+		mxl_ml_build_sta_control_field(buf, lhapd, ML_ASSOC);
 
 		/* STA Info */
-		wpabuf_put_u8(buf, ML_PER_STA_INFO_LEN);
-		wpabuf_put_data(buf, lhapd->own_addr, ETH_ALEN);
-		wpabuf_put_le16(buf, lhapd->iconf->beacon_int);
-#ifdef CONFIG_WIFI_CERTIFICATION
-		if((hapd->iface->current_mode->mode == HOSTAPD_MODE_IEEE80211B ||
-				 hapd->iface->current_mode->mode == HOSTAPD_MODE_IEEE80211G) ||
-				 (lhapd->iface->current_mode->mode == HOSTAPD_MODE_IEEE80211B ||
-				  lhapd->iface->current_mode->mode == HOSTAPD_MODE_IEEE80211G))
-			tsf = hapd->iconf->mxl_conf.eht_mld_tsf_diff;
-#endif /* CONFIG_WIFI_CERTIFICATION */
-		wpabuf_put_le64(buf, tsf);
-		wpabuf_put_u8(buf, 0); /* DTIM Count */
-		wpabuf_put_u8(buf, lhapd->conf->dtim_period);
-		/* Param change count */
-		wpabuf_put_u8(buf, 0);
+		sta_info_len = mxl_ml_build_sta_info(buf, lhapd, hapd, ML_ASSOC);
 
-		if (!per_sta_cap)
-			continue;
+		/* Prepare per-STA Profile subelement length */
+		total_len = LINK_INFO_STA_CTRL_LEN + sta_info_len + per_sta_cap_len;
 
 		/* Fragment the sub element if needed */
-		if (total_len <= 255) {
-			wpabuf_put_data(buf, per_sta_cap,
-					per_sta_cap_len);
+		if (total_len <= MAX_IE_LEN) {
+			*link_info_len = total_len;
+			wpabuf_put_data(buf, per_sta_cap, per_sta_cap_len);
 		} else {
-			ptr = per_sta_cap;
-			len = per_sta_cap_len;
-
-			slice_len = 255 - sta_info_len;
-
-			wpabuf_put_data(buf, ptr, slice_len);
-			len -= slice_len;
-			ptr += slice_len;
-
-			while (len) {
-				if (len <= 255)
-					slice_len = len;
-				else
-					slice_len = 255;
-
-				wpabuf_put_u8(buf, EHT_ML_SUB_ELEM_FRAGMENT);
-				wpabuf_put_u8(buf, slice_len);
-				wpabuf_put_data(buf, ptr, slice_len);
-
-				len -= slice_len;
-				ptr += slice_len;
-			}
+			*link_info_len = MAX_IE_LEN;
+			mxl_ml_fragment_per_sta_profile(buf, per_sta_cap, per_sta_cap_len, total_len, sta_info_len);
 		}
 	}
 
-	/* Fragment the Multi-Link element, if needed */
-	len = wpabuf_len(buf);
-	ptr = wpabuf_head(buf);
-
-	if (len <= 254)
-		slice_len = len;
-	else
-		slice_len = 254;
-
-	*pos++ = WLAN_EID_EXTENSION;
-	*pos++ = slice_len + 1;
-	*pos++ = WLAN_EID_EXT_MULTI_LINK;
-	os_memcpy(pos, ptr, slice_len);
-
-	ptr += slice_len;
-	pos += slice_len;
-	len -= slice_len;
-
-	while (len) {
-		if (len <= 255)
-			slice_len = len;
-		else
-			slice_len = 255;
-
-		*pos++ = WLAN_EID_FRAGMENT;
-		*pos++ = slice_len;
-		os_memcpy(pos, ptr, slice_len);
-
-		ptr += slice_len;
-		pos += slice_len;
-		len -= slice_len;
-	}
+	/* Prepare the Multi-Link IE, fragment if needed */
+	pos = mxl_ml_ie_prepare(buf, pos);
 
 	wpabuf_free(buf);
 	return pos;
 }
 
-u8 * mxl_hostapd_eid_multilink(struct hostapd_data *hapd, struct sta_info *sta, u8 *eid,
-		enum multilink_type ml_type, u16 link_sta_status, bool is_auth, bool is_non_tx_ml_probe)
-{
-	u16 multilink_ctrl = 0;
-	u16 msd_info = 0;
-	u16 eml_capab = 0;
-	u16 mld_capab = 0;
-	u16 sta_ctrl = 0;
-	u16 beacon_intr = 0;
-	u64 tsf = 0;
-	u8 *size_pos = NULL;
-	u8 *link_info_len = NULL;
-	u8 *sta_info_size_pos = NULL;
-	u8 *data_pos = NULL;
-	u8 *link_info_data_pos = NULL;
-	u8 common_info_len;
-	u8 fw_ie_len = 0;
-	struct hostapd_data *sibling_hapd;
-	u8 mld_num_simul_links = mxl_ml_get_num_of_sim_links(hapd);
-	bool mld_mediumsync_present = hapd->conf->mxl_bss_conf.mld_mediumsync_present;
-	u8 bpcc_cnt = hapd->conf->mxl_bss_conf.mld_bpcc_count;
-
-	*eid++ = WLAN_EID_EXTENSION;
-	size_pos = eid;
-
-	/* size not known yet */
-	*eid++ = 0;
-	data_pos = eid;
-
-	*eid++ = WLAN_EID_EXT_MULTI_LINK;
-
-	/* Multi-Link control element */
-	/* B0..B2 */
-#ifdef CONFIG_EPCS_MXL
-	if (ml_type == ML_EPCS)
-		multilink_ctrl |= MULTI_LINK_CONTROL_TYPE_PRIOR_ACCESS;
-	else
-#endif /* CONFIG_EPCS_MXL */
-		multilink_ctrl |= MULTI_LINK_CONTROL_TYPE_BASIC;
-	/* B4, B5, B7, B8 */
-#ifdef CONFIG_EPCS_MXL
-	if ((!is_auth) && (ml_type != ML_EPCS)) {
-#else
-	if (!is_auth) {
-#endif /* CONFIG_EPCS_MXL */
-		multilink_ctrl |= (MLO_LINK_ID_INFO_PRESENT | MLO_BSS_PARAM_CHANGE_COUNT_PRESENT |
-				MLO_MLD_CAPAB_PRESENT);
-		multilink_ctrl |= MLO_EML_CAPAB_PRESENT;
-		if (mld_mediumsync_present && ml_type != ML_BASIC)
-			multilink_ctrl |= MLO_MSDI_PRESENT;
-		if (ml_type == ML_PROBE && is_non_tx_ml_probe) {
-			multilink_ctrl |= MLO_MLD_ID_PRESENT;
-		}
-	}
-
-	WPA_PUT_LE16(eid, multilink_ctrl);
-	eid += sizeof(multilink_ctrl);
-
-	/* Multi-Link Common info */
-#ifdef CONFIG_EPCS_MXL
-	if ((is_auth) || (ml_type == ML_EPCS))
-#else
-	if (is_auth)
-#endif /* CONFIG_EPCS_MXL */
-		common_info_len = COMMON_INFO_LEN_AUTH;
-	else {
-		common_info_len = COMMON_INFO_LEN;
-		if (multilink_ctrl & MLO_MLD_ID_PRESENT)
-			common_info_len += 1;
-		if (mld_mediumsync_present && ml_type != ML_BASIC)
-			common_info_len += MULTILINK_MSD_LEN;
-	}
-	*eid++ = common_info_len;
-
-	memcpy_s(eid, ETH_ALEN, hapd->conf->mxl_bss_conf.ap_mld_mac, ETH_ALEN);
-	eid += ETH_ALEN;
-#ifdef CONFIG_EPCS_MXL
-	if ((!is_auth) && (ml_type != ML_EPCS)) {
-#else
-	if (!is_auth) {
-#endif /* CONFIG_EPCS_MXL */
-		*eid++ =  hapd->conf->mxl_bss_conf.link_id;
-		/* BSS Parameters Change Count */
-		*eid++ = bpcc_cnt;
-		/* Medium Synchronization Delay Information */
-		if (mld_mediumsync_present && ml_type != ML_BASIC) {
-			msd_info = hapd->conf->mxl_bss_conf.mld_mediumsync_duration;
-			msd_info |= hapd->conf->mxl_bss_conf.mld_mediumsync_ofdmedthresh << MLO_MEDIUM_SYNC_OFDMED_THRESH_START_POSITION;
-			msd_info |= hapd->conf->mxl_bss_conf.mld_mediumsync_maxtxop << MLO_MEDIUM_SYNC_MAX_TXOP_START_POSITION;
-			WPA_PUT_LE16(eid, msd_info);
-			eid += sizeof(msd_info);
-		}
-		if (hapd->iconf->mxl_conf.mlo_supported == MULTILINK_EMLSR)
-			eml_capab |= MLO_EMLSR;
-		else if (hapd->iconf->mxl_conf.mlo_supported == MULTILINK_EMLMR)
-			eml_capab |= MLO_EMLMR;
-		eml_capab |= hapd->conf->mxl_bss_conf.eml_transition_timeout << MLO_TRANSITION_TIMEOUT_MASK;/* Temporary change to pass compilation */
-		WPA_PUT_LE16(eid, eml_capab);
-		eid += sizeof(eml_capab);
-
-		mld_capab |= mld_num_simul_links;
-		if (hapd->conf->mxl_bss_conf.mlo_t2lm_support)
-			mld_capab |= MLO_T2LM_NEGOTIATION_SUPPORT_MAP_ALL_TID_TO_SAME_LINK;
-		WPA_PUT_LE16(eid, mld_capab);
-		eid += sizeof(mld_capab);
-		/* MLD ID */
-		if (multilink_ctrl & MLO_MLD_ID_PRESENT)
-			*eid++ = hapd->conf->bss_idx - 1; /* BSSID IDX */
-	}
-
-	wpa_printf(MSG_DEBUG,"MXL_MLO: ml_type %d, MCST add host %d MCST add fw %d",
-			      ml_type, hapd->mxl_data.ml_ie_max_chan_sw_time_ie_add,
-			      hapd->mxl_data.ml_ie_max_chan_sw_time_ie_len_add);
-
-	/* Multi-Link Link Info(Per-STA) */
-	if ((ml_type != ML_BASIC) ||
-	    ((hapd->mxl_data.ml_ie_max_chan_sw_time_ie_add) ||
-	     (hapd->mxl_data.ml_ie_max_chan_sw_time_ie_len_add))) {
-		sibling_hapd = hapd->mxl_data.sibling_hapd[LINK1]; /* TODO: WLANRTSYS-95653 - Need to revisit during the CSA task */
-		if (!sibling_hapd) {
-			wpa_printf(MSG_ERROR,"MXL_MLO: sibling hapd is NULL!!!");
-			return eid;
-		}
-		fw_ie_len = 0;
-		*eid++ = WLAN_EID_SUBELEMENT_MLO_PER_STA_PROFILE;
-		link_info_len = eid;
-		/* size not known yet */
-		*eid++ = 0;
-		link_info_data_pos = eid;
-#ifdef CONFIG_EPCS_MXL
-		if (ml_type == ML_EPCS) {
-			sta_ctrl = hapd->conf->mxl_bss_conf.link_id;
-			sta_ctrl = host_to_le16(sta_ctrl);
-			memcpy_s(eid, sizeof(sta_ctrl), &sta_ctrl, sizeof(sta_ctrl));
-			eid += sizeof(sta_ctrl);
-		} else {
-#endif /* CONFIG_EPCS_MXL */
-		sta_ctrl |= (sibling_hapd->conf->mxl_bss_conf.link_id | MLO_LINK_INFO_COMPLETE_PROFILE | MLO_LINK_INFO_MACADDR_PRESENT |
-				MLO_LINK_INFO_BI_PRESENT | MLO_LINK_INFO_TSF_OFFSET_PRESENT | MLO_LINK_INFO_DTIM_INFO_PRESENT);
-		if (ml_type == ML_ASSOC)
-			sta_ctrl |= (MLO_LINK_INFO_BSS_PARAM_CHANGE_CNT_PRESENT);
-
-		/* CSA Multilink case, beacons to add partial profile with CSA/ECSA/MCST IE */
-		if (ml_type == ML_BASIC)
-			sta_ctrl &= ~MLO_LINK_INFO_COMPLETE_PROFILE;
-
-		WPA_PUT_LE16(eid, sta_ctrl);
-		eid += sizeof(sta_ctrl);
-		/* sta_info */
-		/* sta_info len */
-		sta_info_size_pos = eid;
-		*eid++ = 0;
-		memcpy_s(eid, ETH_ALEN, sibling_hapd->own_addr, ETH_ALEN);
-		eid += ETH_ALEN;
-		beacon_intr = sibling_hapd->iconf->beacon_int;
-		WPA_PUT_LE16(eid, beacon_intr);
-		eid += sizeof(beacon_intr);
-#ifdef CONFIG_WIFI_CERTIFICATION
-		if((hapd->iface->current_mode->mode == HOSTAPD_MODE_IEEE80211B ||
-				 hapd->iface->current_mode->mode == HOSTAPD_MODE_IEEE80211G) ||
-				 (sibling_hapd->iface->current_mode->mode == HOSTAPD_MODE_IEEE80211B ||
-				  sibling_hapd->iface->current_mode->mode == HOSTAPD_MODE_IEEE80211G))
-			tsf = hapd->iconf->mxl_conf.eht_mld_tsf_diff;
-#endif /* CONFIG_WIFI_CERTIFICATION */
-		WPA_PUT_LE64(eid, tsf);
-		eid += sizeof(tsf);
-		WPA_PUT_LE16(eid, (sibling_hapd->conf->dtim_period << 8));
-		eid += sizeof(le16);
-		if (ml_type == ML_ASSOC) {
-			*eid++ = 0;
-		}
-
-		*sta_info_size_pos = eid - sta_info_size_pos;
-#ifdef CONFIG_EPCS_MXL
-		}
-#endif /* CONFIG_EPCS_MXL */
-		/* Per-STA profile */
-		switch (ml_type) {
-			case ML_BASIC:
-				eid = mxl_hostapd_eid_mld_staprof_bcn(hapd, eid, &fw_ie_len);
-				wpa_printf(MSG_INFO,"MXL_MLO: fw ie len %d!!!", fw_ie_len);
-				break;
-			case ML_PROBE:
-				/* TODO
-				 * include the CSA related IEs in the Probe Response
-				 * during a CSA on the affiliated link
-				 */
-				eid = mxl_hostapd_eid_mld_staprof_probe(sibling_hapd, eid); /* TODO: WLANRTSYS-95651 - Need to handle during the ML Probe response task */
-				break;
-#ifdef CONFIG_EPCS_MXL
-			case ML_EPCS:
-				eid = mxl_hostapd_eid_mld_staprof_epcs(hapd, eid);
-				break;
-#endif /* CONFIG_EPCS_MXL */
-			default:
-				wpa_printf(MSG_ERROR,"MXL_MLO: NO valid ML type(%d)!!!", ml_type);
-				return eid;
-		}
-
-		if ((eid - link_info_data_pos) <= MAX_IE_LEN) {
-			/* fw_ie_len is the total length of CSA/CSW/ECSA IEs to be
-			 * added by fw in per STA profile during CSA. Host updates
-			 * the per STA profile length factoring in the fw adding IEs
-			 */
-			*link_info_len = (eid - link_info_data_pos) + (fw_ie_len);
-		} else {
-			*link_info_len = MAX_IE_LEN;
-		}
-
-		/* Check if fragmentation needed */
-		if ((eid - data_pos) > MAX_IE_LEN) {
-			*size_pos = MAX_IE_LEN;
-			eid = mxl_hostapd_eid_mld_frag(data_pos,
-				common_info_len, (eid - link_info_data_pos));
-		}
-	}
-	if (size_pos && (*size_pos == 0)) {
-		*size_pos = (eid - data_pos);
-		wpa_hexdump(MSG_DEBUG, "mxl_hostapd_eid_multilink", data_pos, *size_pos);
-	}
-	return eid;
-}
-
 static void mxl_mlo_set_assoc_link_id(struct sta_info *sta, u8 assoc_link_id)
 {
 	struct sta_info *lsta = NULL;
@@ -1827,7 +1770,7 @@ static int mxl_mld_set_linked_stas(struct sta_info *sta)
 	return 0;
 }
 
-static u16 mxl_mld_proccess_per_sta_profile(struct hostapd_data *hapd, struct sta_info *sta,
+static u16 mxl_mld_process_per_sta_profile(struct hostapd_data *hapd, struct sta_info *sta,
 											   struct wpabuf *mlbuf, const u8 *per_sta_ie)
 {
 	const u8 *pos, *end;
@@ -2125,7 +2068,7 @@ u16 mxl_multilink_ie_validate(struct hostapd_data *hapd, struct sta_info *sta,
 	}
 
 	/* Check and handle per-STA Profiles */
-	status = mxl_mld_proccess_per_sta_profile(hapd, sta, mlbuf, per_sta_ie_start);
+	status = mxl_mld_process_per_sta_profile(hapd, sta, mlbuf, per_sta_ie_start);
 	if (status != WLAN_STATUS_SUCCESS) {
 		wpa_printf(MSG_ERROR, "MXL_MLO: Failed to process per-STA profile");
 		goto out;
diff --git a/src/ap/mxl_mld.h b/src/ap/mxl_mld.h
index 4bf3da10c..596f952c9 100644
--- a/src/ap/mxl_mld.h
+++ b/src/ap/mxl_mld.h
@@ -32,7 +32,7 @@
 /* Len + MLD-MAC + LINK_ID + EML_CAP + MLD_CAP + BPCC*/
 #define COMMON_INFO_LEN                 13
 /* Len + MLD-MAC */
-#define COMMON_INFO_LEN_AUTH            7
+#define COMMON_INFO_LEN_AUTH            (1 + ETH_ALEN)
 /* len + MULTILINK_CTRL_LEN + COMMON_INFO_LEN */
 #define BASIC_MULTILINK_IE_TAG_LEN      (1 + MULTILINK_CTRL_LEN + COMMON_INFO_LEN)
 #define BASIC_MULTILINK_IE_LEN (MULTILINK_IE_HDR_LEN + MULTILINK_CTRL_LEN + COMMON_INFO_LEN)
@@ -272,9 +272,8 @@ int mxl_ml_deinit_link_free(struct hostapd_data *hapd);
 int mxl_hostapd_multilink_ie_len(struct hostapd_data *hapd, enum multilink_type ml_type);
 
 /* Multi-Link Element */
-u8 * mxl_hostapd_eid_multilink(struct hostapd_data *hapd, struct sta_info *sta,
-				u8 *eid, enum multilink_type ml_type, u16 link_sta_status, bool is_auth,
-				bool is_non_tx_ml_probe);
+u8 * mxl_hostapd_eid_multilink_common(struct hostapd_data *hapd, u8 *eid,
+								bool include_mld_id, enum multilink_type ml_type);
 
 /* Multi-Link Element for Association */
 u8 * mxl_ml_eid_multilink_assoc(struct hostapd_data *hapd, struct sta_info *sta, u8 *eid);
diff --git a/src/ap/mxl_wmm.c b/src/ap/mxl_wmm.c
index 36f90bb48..34b6182c7 100644
--- a/src/ap/mxl_wmm.c
+++ b/src/ap/mxl_wmm.c
@@ -123,8 +123,7 @@ static void mxl_handle_epcs_req(struct hostapd_data *hapd,
 	mgmt_resp->u.action.u.eht_epcs_resp.dialog_token = mgmt->u.action.u.eht_epcs_req.dialog_token;
 	mgmt_resp->u.action.u.eht_epcs_resp.status = WLAN_STATUS_SUCCESS;
 
-	pos = mxl_hostapd_eid_multilink(hapd, NULL, mgmt_resp->u.action.u.eht_epcs_resp.variable,
-		ML_EPCS, 0, false, false);
+	pos = mxl_hostapd_eid_multilink_common(hapd, mgmt_resp->u.action.u.eht_epcs_resp.variable, false, ML_EPCS);
 	len = pos - buf;
 
 	if (len != EPCS_RESPONSE_FRAME_LENGTH) {
-- 
2.43.0

